lib/spack/spack/relocate.py

```diff
@@ -0,0 +1,273 @@
+                idpath = rhs
+            if lhs == 'name' and last_cmd == 'LC_LOAD_DYLIB':
+                deps.append(rhs)
+    return rpaths, deps, idpath
+
+
+def macho_make_paths_relative(path_name, old_dir, rpaths, deps, idpath):
+    id = None
+    nrpaths = []
+    ndeps = []
+    if idpath:
+        id = '@rpath/%s' % os.path.basename(idpath)
+    for rpath in rpaths:
+        if re.match(old_dir, rpath):
+            rel = os.path.relpath(rpath, start=os.path.dirname(path_name))
+            nrpaths.append('@loader_path/%s' % rel)
+        else:
+            nrpaths.append(rpath)
+    for dep in deps:
+        if re.match(old_dir, dep):
+            rel = os.path.relpath(dep, start=os.path.dirname(path_name))
+            ndeps.append('@loader_path/%s' % rel)
+        else:
+            ndeps.append(dep)
+    return nrpaths, ndeps, id
+
+
+def macho_replace_paths(old_dir, new_dir, rpaths, deps, idpath):
+    id = None
+    nrpaths = []
+    ndeps = []
+    if idpath:
+        id = idpath.replace(old_dir, new_dir)
+    for rpath in rpaths:
+        nrpath = rpath.replace(old_dir, new_dir)
+        nrpaths.append(nrpath)
+    for dep in deps:
+        ndep = dep.replace(old_dir, new_dir)
+        ndeps.append(ndep)
+    return nrpaths, ndeps, id
+
+
+def modify_macho_object(path_name, old_dir, new_dir, relative):
+    """
+    Modify MachO binaries by changing rpaths,and id and dependency lib paths.
+    Examines the output of otool -l for these three fields:
+    LC_ID_DYLIB, LC_LOAD_DYLIB, LC_RPATH
+    The old install dir in LC_ID_DYLIB is replaced with the new install dir
+    using install_name_tool -id newid binary
+    The old install dir in LC_LOAD_DYLIB is replaced with the new install dir
+    using install_name_tool -change old new binary
+    The old install dir in LC_RPATH is replaced with the new install dir using
+    install_name_tool  -rpath old new binary
+    """
+    # avoid error message for libgcc_s
+    if 'libgcc_' in path_name:
+        return
+    rpaths, deps, idpath = macho_get_paths(path_name)
+    id = None
+    nrpaths = []
+    ndeps = []
+    if relative:
+        nrpaths, ndeps, id = macho_make_paths_relative(path_name,
+                                                       old_dir, rpaths,
+                                                       deps, idpath)
+    else:
+        nrpaths, ndeps, id = macho_replace_paths(old_dir, new_dir, rpaths,
+                                                 deps, idpath)
+    st = os.stat(path_name)
+    # some installs create read-only binaries
+    install_name_tool = Executable('install_name_tool')
+    if id:
+        install_name_tool('-id', id, path_name, output=str, err=str)
+
+    for orig, new in zip(deps, ndeps):
+        install_name_tool('-change', orig, new, path_name)
+
+    for orig, new in zip(rpaths, nrpaths):
+        install_name_tool('-rpath', orig, new, path_name)
+        os.chmod(path_name, st.st_mode)
```

Comment:
Does this no longer apply? I don't see you changing ```st_mode``` anymore (which I like). In that case the ```os.stat``` on line 162 is no longer required either.